doc/devguide/extending/app-layer/app-layer-frames.rst

```diff
@@ -0,0 +1,129 @@
+
+The goal of the stream *Frame* is to expose application layer protocol `PDUs <https://en.wikipedia.org/wiki/Protocol_data_unit>`_ and other such arbitrary elements to the detection engine directly, instead of relying on Transactions. The main purpose is to bring *TCP.data* processing times down by specialising/ filtering down traffic detection.
+
+Adding Frame Support to a Parser
+================================
+
+The application layer parser exposes frames it supports to the detect engine, by tagging them as they're parsed. The rest works automatically.
+
+In order to allow the engine to identify frames for records of a given application layer parser, you'll have to decide which frames make sense for the specific protocol you are handling. Once you have that, the basic steps are:
+
+- create an enum for the frame types;
+- identify the parsing function(s) where application layer records are parsed;
+- identify the correct moment to use the calls;
+- use the Frame API calls directly or build upon them to register the frames.
+
+Once this is done, you can enable frame eve-output to confirm that your frames are being registered correctly. If done correctly, you should be able to write a rule using the *frame* keyword and the frame type you have registered with the application layer parser.
+
+Implementation & API Callbacks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The frame types are defined as an enum. In Rust, derive from the ``AppLayerFrameType``:
+
+.. literalinclude:: ../../../../rust/src/smb/smb.rs
+    :caption: rust/src/smb/smb.rs
+    :language: rust
+    :start-at: #[derive(AppLayerFrameType)]
+    :lines: 1-15
+
+How some frames are registered in the `SMB <https://github.com/OISF/suricata/blob/master/rust/src/smb/smb.rs#L1383>`_ parser:
+
+.. literalinclude:: ../../../../rust/src/smb/smb.rs
+    :caption: rust/src/smb/smb.rs
+    :language: rust
+    :start-at: pub fn parse_tcp_data_ts<'b>
+    :end-at: nbss_hdr.data, nbss_hdr.length as i64);
+    :lines: 1-3, 55-74
+    :dedent: 4
+
+The function calls for registering ``frames`` are available from *frames.rs*. A frame to the server could be registered like so:
+
+.. literalinclude:: ../../../../rust/src/frames.rs
+    :caption: rust/src/frames.rs
+    :language: rust
+    :start-at: pub fn new_ts
+    :end-before: pub fn new_tc
+    :dedent: 4
+
+And to the client:
+
+.. literalinclude:: ../../../../rust/src/frames.rs
+    :caption: rust/src/frames.rs
+    :language: rust
+    :start-at: pub fn new_tc
+    :lines: 1-6
+    :dedent: 4
+
+The parameters represent:
+
+- ``flow``: dedicated data type to carry specific flow-related data
+- ``stream_slice``: dedicated data type to carry stream data, shown bellow
+- ``frame_start``: a pointer to the start of the frame buffer in the stream, ``cur_i`` in the example code snippet
+- ``frame_len``: what we expect the frame length to be (the engine may need to wait until it has enough data...)
```

Comment:
I think this may require a bit more explanation. Maybe telnet example?

Reply:
"if the length is unknown pass -1"
"length can exceed the current input"

In telnet or other protocols that search for a end of frame char we'd generally start with length unknown, then update the length when the char is found

In smb and other binary protocols where the frames have explicit length fields we set frames to match those lengths, even if the length is bigger than the current input